main.go

```diff
@@ -50,53 +70,99 @@ func init() {
-	flag.DurationVar(&syncPeriod, "sync-period", 10*time.Minute,
+
+	fs.StringVar(&watchNamespace, "namespace", "",
+		"Namespace that the controller watches to reconcile cluster-api objects. If unspecified, the controller watches for cluster-api objects across all namespaces.")
+
+	fs.StringVar(&watchFilterValue, "watch-filter", "",
+		fmt.Sprintf("Label value that the controller watches to reconcile cluster-api objects. Label key is always %s. If unspecified, the controller watches for all cluster-api objects.", clusterv1.WatchLabel))
+
+	fs.StringVar(&profilerAddress, "profiler-address", "",
+		"Bind address to expose the pprof profiler (e.g. localhost:6060)")
+
+	fs.IntVar(&helmChartProxyConcurrency, "helm-chart-proxy-concurrency", 10,
+		"The number of HelmChartProxies to process concurrently.")
+
+	fs.IntVar(&helmReleaseProxyConcurrency, "helm-release-proxy-concurrency", 10,
+		"The number of HelmReleaseProxies to process concurrently.")
+
+	fs.DurationVar(&syncPeriod, "sync-period", 10*time.Minute,
 		"The minimum interval at which watched resources are reconciled (e.g. 15m)")
+
+	fs.IntVar(&webhookPort, "webhook-port", 9443,
+		"Webhook Server port")
+
+	fs.StringVar(&webhookCertDir, "webhook-cert-dir", "/tmp/k8s-webhook-server/serving-certs/",
+		"Webhook cert dir, only used when webhook-port is specified.")
+
+	fs.StringVar(&healthAddr, "health-addr", ":9440",
+		"The address the health endpoint binds to.")
+
+	flags.AddDiagnosticsOptions(fs, &diagnosticsOptions)
+
+	feature.MutableGates.AddFlag(fs)
+}
+
+// Add RBAC for the authorized diagnostics endpoint.
+// +kubebuilder:rbac:groups=authentication.k8s.io,resources=tokenreviews,verbs=create
+// +kubebuilder:rbac:groups=authorization.k8s.io,resources=subjectaccessreviews,verbs=create
+
+func main() {
+	InitFlags(pflag.CommandLine)
+	klog.InitFlags(flag.CommandLine)
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
 	// Set log level 2 as default.
-	if err := flag.Set("v", "2"); err != nil {
+	if err := pflag.CommandLine.Set("v", "2"); err != nil {
 		setupLog.Error(err, "failed to set log level: %v")
 		os.Exit(1)
 	}
-	flag.Parse()
+	pflag.Parse()
+
+	diagnosticsOpts := flags.GetDiagnosticsOptions(diagnosticsOptions)
+
+	var watchNamespaces map[string]cache.Config
+	if watchNamespace != "" {
+		watchNamespaces = map[string]cache.Config{
+			watchNamespace: {},
+		}
+	}
+
+	// klog.Background will automatically use the right logger.
+	ctrl.SetLogger(klog.Background())
 
-	ctrl.SetLogger(klogr.NewWithOptions(klogr.WithFormat(klogr.FormatKlog)))
 	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
 		Scheme:                 scheme,
-		MetricsBindAddress:     metricsAddr,
-		Port:                   9443,
-		HealthProbeBindAddress: probeAddr,
 		LeaderElection:         enableLeaderElection,
-		LeaderElectionID:       "5a2dee3e.cluster.x-k8s.io",
+		LeaderElectionID:       "controller-leader-election-caaph",
```

Comment:
This seems better. :-) Is there any convention for `LeaderElectionID` in other controllers? (Should it include a namespace like `cluster.x-k8s.io`?)

Reply:
I copied this from CAPZ, which had it as `controller-leader-election-capz`. Slowly working through the copy/paste issues!